okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/QosExceptionResponseHandler.java

```diff
@@ -16,74 +16,19 @@
+import com.palantir.conjure.java.QosExceptionResponseMapper;
 import com.palantir.conjure.java.api.errors.QosException;
-import com.palantir.logsafe.SafeArg;
-import com.palantir.logsafe.UnsafeArg;
-import java.net.MalformedURLException;
-import java.net.URL;
-import java.time.Duration;
 import java.util.Optional;
 import okhttp3.Response;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 /**
  * A {@link ResponseHandler} that turns QOS-related HTTP responses into {@link QosException}s.
  */
 enum QosExceptionResponseHandler implements ResponseHandler<QosException> {
     INSTANCE;
 
-    private static final Logger log = LoggerFactory.getLogger(QosExceptionResponseHandler.class);
-
     @Override
     public Optional<QosException> handle(Response response) {
-        switch (response.code()) {
-            case 308:
-                return handle308(response);
-            case 429:
-                return Optional.of(handle429(response));
-            case 503:
-                return Optional.of(handle503());
-        }
-
-        return Optional.empty();
-    }
-
-    private Optional<QosException> handle308(Response response) {
-        String locationHeader = response.header(HttpHeaders.LOCATION);
-        if (locationHeader == null) {
-            log.error("Retrieved HTTP status code 308 without Location header, cannot perform "
-                    + "redirect. This appears to be a server-side protocol violation.");
-            return Optional.empty();
-        }
-        // Note: Do not SafeArg-log the redirectTo URL since it typically contains unsafe information
-        log.debug("Received 308 response, retrying host at advertised location",
-                SafeArg.of("location", locationHeader));
-
-        try {
-            return Optional.of(QosException.retryOther(new URL(locationHeader)));
-        } catch (MalformedURLException e) {
-            log.error("Failed to parse location header, not performing redirect",
-                    UnsafeArg.of("locationHeader", locationHeader), e);
-            return Optional.empty();
-        }
-    }
-
-    private static QosException handle429(Response response) {
-        String duration = response.header(HttpHeaders.RETRY_AFTER);
-        if (duration == null) {
-            log.debug("Received 429 response, throwing QosException to trigger delayed retry");
-            return QosException.throttle();
-        } else {
-            log.debug("Received 429 response, throwing QosException to trigger delayed retry",
-                    SafeArg.of("duration", duration));
-            return QosException.throttle(Duration.ofSeconds(Long.parseLong(duration)));
-        }
-    }
-
-    private static QosException handle503() {
-        log.debug("Received 503 response, throwing QosException to trigger failover");
-        return QosException.unavailable();
+        return QosExceptionResponseMapper.mapResponseCode(response.code(), response::header);
```

Comment:
Let me double check what you're asking for here. Would you like `feignimpl.QosErrorDecoder` and `retrofit2.QosExceptionThrowingCallAdapterFactory` to depend directly on `okhttp.QosExceptionResponseHandler.handle`? I created the helper function because `feignimpl.QosErrorDecoder` receives a `feign.Response` and retrofit2.QosExceptionThrowingCallAdapterFactory` receives a `retrofit2.Response` whereas `okhttp.QosExceptionResponseHandler.handle` receives a `okhttp3.Response`.

Reply:
I think @pkoenig10 did not notice that `QosExceptionResponseMapper` was also called from `QosExceptionThrowingCallAdapterFactory` and `QosErrorDecoder`; current way looks right to me